### PR TITLE
#516 - Discretize using matrix interval remainder

### LINKED ISSUE #516: Discretize using matrix interval remainder
The approach in [Proposition 3.1, Section 3, 1] is to use an approximation model that applies an appropriate correction matrix `F`, an interval matrix, applied to the set of initial states.

This method can be used in general. Moreover, it is particularly interesting when:

- zonotopes are used as set representation, because one can overapproximate again with a zonotope
- the model has uncertain parameters

Related to this issue: [IntervalMatrices.jl](https://github.com/JuliaReach/IntervalMatrices.jl).

[1]  Althoff, Matthias. Reachability analysis and its application to the safety
assessment of autonomous cars. Diss. Technische Universität München, 2010.

### PR DESCRIPTION
**Pull request by schillic**
Closes #516.

* [x] homogeneous case
  * [x] lazy implementation
  * [x] zonotope implementation
* [x] inhomogeneous case

### CODE REVIEW

File: src/ReachSets/discretize.jl [hunk 1]
```diff
@@ -1,3 +1,6 @@
+using IntervalMatrices: expm_overapproximation, correction_hull, _expm_remainder
+IM = IntervalMatrices
```
Comment (mforets) [inline src/ReachSets/discretize.jl]:
```suggestion
const IM = IntervalMatrices
```

File: src/ReachSets/discretize.jl [hunk 2]
```diff
@@ -720,3 +739,154 @@ function _discretize_interpolation_inhomog(δ, U0, U, X0, ϕ, Einit, Eψ0, A, si
...
+    end
+
+    Ω0, Ud = _discretize_interval_matrix_inhomog(U, Ω0_homog, linear_maps,
+                                                 set_ops)
+
+    # create identity interval matrix
+    one_interval = IntervalMatrices.Interval(one(N), one(N))
+    B = IntervalMatrix(fill(zero_interval, (n, n)))
+    @inbounds for i in 1:n
+        B[i, i] = one_interval
```
Comment (mforets) [inline src/ReachSets/discretize.jl]: This calls for making a new function in `IntervalMatrices` for that purpose.

File: src/ReachSets/discretize.jl [hunk 3]
```diff
@@ -720,3 +739,154 @@ function _discretize_interpolation_inhomog(δ, U0, U, X0, ϕ, Einit, Eψ0, A, si
...
+    if inputdim(𝑆) == 0
+        return IVP(CLDS(ϕ, stateset(𝑆.s)), Ω0_homog)
+    end
+
+    U = inputset(𝑆)
+    U0 = next_set(U, 1)
+    n = size(A, 1)
+    linear_maps = Vector{LinearMap{N}}(undef, order > 2 ? 3 : 2)
+
+    IδW = 1/2 * δ^2 * A + 1/6 * δ^3 * A * A
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/ReachSets/discretize.jl]: I agree that the paper uses the exact computation, but I do not see how you can use the linked function for that. The second fraction is `1/6`, which, using your factor `δ/2`, would still be `1/3`, but the function returns `1/2`.

File: src/ReachSets/discretize.jl [hunk 4]
```diff
@@ -720,3 +739,154 @@ function _discretize_interpolation_inhomog(δ, U0, U, X0, ϕ, Einit, Eψ0, A, si
...
+    @inbounds for i in 1:n
+        IδW[i, i] *= δ
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/ReachSets/discretize.jl]: Agreed, it should be `+`.

File: src/ReachSets/discretize.jl [hunk 5]
```diff
@@ -720,3 +739,154 @@ function _discretize_interpolation_inhomog(δ, U0, U, X0, ϕ, Einit, Eψ0, A, si
...
+    end
+    linear_maps[1] = LinearMap(IδW, U0)
+
+    linear_maps[2] = LinearMap(_expm_remainder(A, δ, order; n=n) * δ, U0)
+
+    zero_interval = IntervalMatrices.Interval(zero(N), zero(N))
+    if order > 2
+        fac_i_plus_1 = 6
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/ReachSets/discretize.jl]: No, the first denominator is `3!*4 = 24`. I initialize the variable to `3! = 6` and then in each iteration multiply by `i+1` (i.e., the first time I multiply by `4`).

File: src/ReachSets/discretize.jl [hunk 6]
```diff
@@ -720,3 +739,154 @@ function _discretize_interpolation_inhomog(δ, U0, U, X0, ϕ, Einit, Eψ0, A, si
...
+        Ai = A * A
+        δ_iplus1 = δ^3
+        M_sum = IntervalMatrix(fill(zero_interval, size(A)))
+        @inbounds for i in 3:order
+            fac_i_plus_1 *= i+1
+            δ_iplus1 *= δ
+            Ai *= A
+            1/fac_i_plus_1 * δ_iplus1 * Ai
+        end
+        linear_maps[3] = LinearMap(M_sum, U0)
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/ReachSets/discretize.jl]: Indeed :smile:
The last line in the loop computes the summand but does not use it.

File: src/ReachSets/discretize.jl (hunk 2 above)
Comment (schillic) [reply src/ReachSets/discretize.jl]: https://github.com/JuliaReach/IntervalMatrices.jl/issues/49

**Comment by schillic** [comment]
Can you rebase please? I cannot read these changes.

File: src/ReachSets/discretize.jl [hunk 7]
```diff
@@ -800,31 +800,30 @@ function discretize_interval_matrix(𝑆::InitialValueProblem, δ::Float64,
...
         M_sum = IntervalMatrix(fill(zero_interval, size(A)))
         @inbounds for i in 3:order
-            fac_i_plus_1 *= i+1
-            δ_iplus1 *= δ
-            Ai *= A
-            1/fac_i_plus_1 * δ_iplus1 * Ai
+            αᵢ₊₁ *= i+1
+            δⁱ⁺¹ *= δ
+            Aⁱ *= A
+            M_sum += 1/αᵢ₊₁ * Aⁱ * δⁱ⁺¹
```
Comment (schillic) [inline src/ReachSets/discretize.jl]: Before I first did the scalar computation and then the matrix scaling. Now your code needs to do two matrix operations. I suspect that the previous version was faster. Was it wrong?

Comment (mforets) [reply src/ReachSets/discretize.jl]: Indeed, i missed that point. Other computations should dominate, but i agree it is better to add parens over the scalar terms.

**Comment by schillic** [comment]
Thanks! :+1: (I cannot approve my own PR.)

**Pull request merged**